### PR TITLE
Decimal functions - fixups visible.

### PR DESCRIPTION
**Pull request by fiedukow**
For @ilfrin

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/type/LongDecimalType.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
 import static com.facebook.presto.spi.type.Decimals.SIZE_OF_LONG_DECIMAL;
 import static com.facebook.presto.spi.type.Decimals.decodeUnscaledValue;
 
-final class LongDecimalType
+public final class LongDecimalType
```
Comment (ilfrin) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/LongDecimalType.java]: need to keep this non-public

File: presto-spi/src/main/java/com/facebook/presto/spi/type/ShortDecimalType.java [hunk 2]
```diff
@@ -23,7 +23,7 @@
 import static com.facebook.presto.spi.type.Decimals.MAX_SHORT_PRECISION;
 import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
 
-final class ShortDecimalType
+public final class ShortDecimalType
```
Comment (ilfrin) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/ShortDecimalType.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java [hunk 3]
```diff
@@ -160,8 +161,8 @@ public static void outputLongDecimal(LongDecimalType type, BigIntegerState state
         }
         else {
             BigDecimal sum = new BigDecimal(state.getBigInteger(), type.getScale());
-            checkOverflow(sum);
-            type.writeBigDecimal(out, sum);
+            checkOverflow(state.getBigInteger());
```
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java:163]: why?

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 4]
```diff
@@ -512,8 +512,8 @@ public void testRound()
         assertFunction("round(DECIMAL '-123456789012345678.501', 1)", createDecimalType(22, 3), SqlDecimal.of("-123456789012345678.500"));
         assertFunction("round(DECIMAL '999999999999999999.99', 1)", createDecimalType(21, 2), SqlDecimal.of("1000000000000000000.00"));
         assertFunction("round(DECIMAL '-999999999999999999.99', 1)", createDecimalType(21, 2), SqlDecimal.of("-1000000000000000000.00"));
-        assertFunction("round(DECIMAL '0.00', 1)", createDecimalType(4, 2), SqlDecimal.of("0.00"));
-        assertFunction("round(DECIMAL '000000000000000000.00', 1)", createDecimalType(21, 2), SqlDecimal.of("0.00"));
+        assertFunction("round(DECIMAL '0.00', 1)", createDecimalType(3, 2), SqlDecimal.of("0.00"));
```
Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:514]: why did this change?

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 5]
```diff
@@ -277,7 +277,7 @@ public void testMod()
         assertFunction("mod(5.0, NULL)", DOUBLE, null);
         assertFunction("mod(NULL, 5.0)", DOUBLE, null);
 
-        assertFunction("mod(DECIMAL '0.0', DECIMAL '2.0')", createDecimalType(2, 1), SqlDecimal.of("0.0"));
+        assertFunction("mod(DECIMAL '0.0', DECIMAL '2.0')", createDecimalType(1, 1), SqlDecimal.of("0.0"));
```
Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:279]: why?

File: presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java [hunk 6]
```diff
@@ -463,29 +463,9 @@ public static SqlScalarFunction modulusScalarFunction(Signature signature)
     {
         return SqlScalarFunction.builder(DecimalOperators.class)
                 .signature(signature)
-                .methods("modulusShortShortShort", "modulusLongLongLong", "modulusShortLongLong", "modulusShortLongShort", "modulusLongShortShort", "modulusLongShortLong")
-                .extraParameters(DecimalOperators::longRescaleExtraParameters)
-                .build();
-    }
-
-    private static SqlScalarFunction decimalModulusOperator()
```
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java]: how come there were two methods with the same signature here?

Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java]: It's due to conflicts resolution - will not be visible after rebase.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/LongDecimalType.java (hunk 1 above)
Comment (fiedukow) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/LongDecimalType.java]: to be done.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/ShortDecimalType.java (hunk 2 above)
Comment (fiedukow) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/ShortDecimalType.java]: to be done.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 4 above)
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:514]: We have now mechanism of cutting precision for output DECIMAL.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 5 above)
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:279]: We have now mechanism of cutting precision for output DECIMAL.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java (hunk 3 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java:163]: It's basically what was done here before.
IDK why the "sum" is called "sum" though - will rename that to "value" to be less confusing ;-)

Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java:163]: ok

File: presto-spi/src/main/java/com/facebook/presto/spi/type/Decimals.java [hunk 7]
```diff
@@ -248,6 +248,11 @@ public static void writeBigDecimal(DecimalType decimalType, BlockBuilder blockBu
         decimalType.writeSlice(blockBuilder, encodeScaledValue(value));
     }
 
+    public static void writeShortDecimal(BlockBuilder blockBuilder, long value)
+    {
+        blockBuilder.writeLong(value).closeEntry();
+    }
+
```
Comment (ilfrin) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/Decimals.java:255]: this name is inconsistent with writeBigDecimal, the latter should be changed, but this is out of scope of this PR

**Comment by ilfrin** [comment]
lgtm

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java (hunk 3 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java:163]: leave the "sum" here, it's a good name, this IS a sum.

**Comment by fiedukow** [comment]
merged to sprint-27
